### DISCUSSION TITLE
Multi GPU Training

### DISCUSSION THREAD

**Discussion by koayst** [coqui-ai/TTS#1819] [General Q&A]
Hi,

I am following the instruction [here](https://tts.readthedocs.io/en/latest/training_a_model.html) to perform the multi-GPU training.

The training will go on until the end of first epoch, an error message will be printed 

 > BEST MODEL : /home/Downloads/TTS/recipes/ljspeech/glow_tts/run-August-03-2022_11+31AM-903a77c1/best_model_203.pth
 ! Run is removed from /home/Downloads/TTS/recipes/ljspeech/glow_tts/run-August-03-2022_11+31AM-903a77c1

The training cannot continue since the "Run" is removed.

Anybody has managed to get multi-gpu training working ?

**Comment by harmlessman** [comment]
I have received a similar warning.
In my experience, such an error occurred when learning was not progressing due to cuda memory issues.
Can you post the full text of the error? If it's the same error I've experienced, I can help.

**Reply by koayst** [reply]
Please give me 2 - 3 days as the machine's GPUs are now being used by another task.  I will reproduce the error and post it here.  Thanks for the help.

**Reply by koayst** [reply]
Hi @harmlessman, @erogol 

Finally, the GPUs are free up for me to conduct the training again.  Here is the entire screen capture.

I have trained align_tts/train-aligntts.py using 2 GPUs and I could finish the training successfully.  The command was:
 CUDA_VISIBLE_DEVICES="2,3" python -m trainer.distribute --script train_aligntts.py

But when I did the same for "glow_tts", somehow the best_model.pth file was deleted or not saved after trainig, hence the error at validation stage.  Similar command was used: CUDA_VISIBLE_DEVICES="2,3" python -m trainer.distribute --script train_glow_tts.py

The error:
FileNotFoundError: /home/xerox/Downloads/TTS/recipes/ljspeech/glow_tts/run-August-15-2022_10+43AM-903a77c1/best_model_203.pth

-------------------------

(coqui) xerox@SOIDGXPPDEV01:~/Downloads/TTS/recipes/ljspeech/glow_tts$ CUDA_VISIBLE_DEVICES="2,3" python -m trainer.distribute --script train_glowtts.py
['train_glowtts.py', '--continue_path=', '--restore_path=', '--group_id=group_2022_08_15-104314', '--use_ddp=true', '--rank=0']
['train_glowtts.py', '--continue_path=', '--restore_path=', '--group_id=group_2022_08_15-104314', '--use_ddp=true', '--rank=1']
 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > log_func:np.log10
 | > min_level_db:-100
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:20
 | > fft_size:1024
 | > power:1.5
 | > preemphasis:0.0
 | > griffin_lim_iters:60
 | > signal_norm:True
 | > symmetric_norm:True
 | > mel_fmin:0
 | > mel_fmax:None
 | > pitch_fmin:0.0
 | > pitch_fmax:640.0
 | > spec_gain:20.0
 | > stft_pad_mode:reflect
 | > max_norm:4.0
 | > clip_norm:True
 | > do_trim_silence:True
 | > trim_db:45
 | > do_sound_norm:False
 | > do_amp_to_db_linear:True
 | > do_amp_to_db_mel:True
 | > do_rms_norm:False
 | > db_level:None
 | > stats_path:None
 | > base:10
 | > hop_length:256
 | > win_length:1024
 | > Found 13100 files in /home/xerox/Downloads/TTS/recipes/ljspeech/LJSpeech-1.1
 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > log_func:np.log10
 | > min_level_db:-100
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:20
 | > fft_size:1024
 | > power:1.5
 | > preemphasis:0.0
 | > griffin_lim_iters:60
 | > signal_norm:True
 | > symmetric_norm:True
 | > mel_fmin:0
 | > mel_fmax:None
 | > pitch_fmin:0.0
 | > pitch_fmax:640.0
 | > spec_gain:20.0
 | > stft_pad_mode:reflect
 | > max_norm:4.0
 | > clip_norm:True
 | > do_trim_silence:True
 | > trim_db:45
 | > do_sound_norm:False
 | > do_amp_to_db_linear:True
 | > do_amp_to_db_mel:True
 | > do_rms_norm:False
 | > db_level:None
 | > stats_path:None
 | > base:10
 | > hop_length:256
 | > win_length:1024
 | > Found 13100 files in /home/xerox/Downloads/TTS/recipes/ljspeech/LJSpeech-1.1

 > Model has 28610257 parameters
config.eval_split_max_size = None
config.eval_split_size = 0.01
config.eval_split_max_size = None
config.eval_split_size = 0.01

 > EPOCH: 0/1000
 --> /home/xerox/Downloads/TTS/recipes/ljspeech/glow_tts/run-August-15-2022_10+43AM-903a77c1
[*] Pre-computing phonemes...


> DataLoader initialization
| > Tokenizer:
        | > add_blank: False
        | > use_eos_bos: False
        | > use_phonemes: True
        | > phonemizer:
                | > phoneme language: en-us
                | > phoneme backend: gruut
| > Number of instances : 12969
  0%|                                                                                                                      | 2/12969 [00:00<26:21,  8.20it/s]ɪnstɛd əv weɪtɪŋ ðɛɹ, ɔzwɔld əpɛɹəntli wɛnt æz fɑɹ əweɪ æz hi kʊd ænd bɔɹdɪd ðə fɚst oʊk klɪf bʌs wɪt͡ʃ keɪm əlɔŋ
 [!] Character '͡' not found in the vocabulary. Discarding it.
 16%|██████████████████▏                                                                                                | 2054/12969 [01:22<05:33, 32.72it/s]ɪntu ðə “kɹeɪtɚ” dʌɡ aʊt ɪn ðə mɪdəl, pɔɹ ðə spʌnd͡ʒ, wɔɹm wɔtɚ, ðə məlæsɪz, ænd soʊdə dɪzɑlvd ɪn hɑt wɔtɚ.
 [!] Character '“' not found in the vocabulary. Discarding it.
ɪntu ðə “kɹeɪtɚ” dʌɡ aʊt ɪn ðə mɪdəl, pɔɹ ðə spʌnd͡ʒ, wɔɹm wɔtɚ, ðə məlæsɪz, ænd soʊdə dɪzɑlvd ɪn hɑt wɔtɚ.
 [!] Character '”' not found in the vocabulary. Discarding it.
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 12969/12969 [05:21<00:00, 40.29it/s]


> DataLoader initialization
| > Tokenizer:
        | > add_blank: False
        | > use_eos_bos: False
        | > use_phonemes: True
        | > phonemizer:
                | > phoneme language: en-us
                | > phoneme backend: gruut
        | > 3 not found characters:
        | > ͡
        | > “
        | > ”
| > Number of instances : 12969
 | > Preprocessing samples
 | > Preprocessing samples
 | > Max text length: 188
 | > Max text length: 188
 | > Min text length: 13
 | > Min text length: 13
 | > Avg text length: 100.90014650319993
 |
 | > Avg text length: 100.90014650319993
 |
 | > Max audio length: 222643.0
 | > Max audio length: 222643.0
 | > Min audio length: 24499.0
 | > Min audio length: 24499.0
 | > Avg audio length: 144984.29755570978
 | > Num. instances discarded samples: 0
 | > Batch group size: 0.
 | > Avg audio length: 144984.29755570978
 | > Num. instances discarded samples: 0
 | > Batch group size: 0.

 > TRAINING (2022-08-15 10:48:55)

   --> STEP: 0/203 -- GLOBAL_STEP: 0
     | > current_lr: 0.00000
     | > step_time: 7.44800  (7.44801)
     | > loader_time: 7.37310  (7.37309)


   --> STEP: 25/203 -- GLOBAL_STEP: 25
     | > loss: 3.68468  (3.68637)
     | > log_mle: 0.76222  (0.75664)
     | > loss_dur: 2.92246  (2.92973)
     | > amp_scaler: 16384.00000  (16384.00000)
     | > grad_norm: 10.68828  (10.02391)
     | > current_lr: 0.00000
     | > step_time: 1.18540  (1.40978)
     | > loader_time: 0.02560  (0.02045)


   --> STEP: 50/203 -- GLOBAL_STEP: 50
     | > loss: 3.70598  (3.68196)
     | > log_mle: 0.74496  (0.75612)
     | > loss_dur: 2.96103  (2.92584)
     | > amp_scaler: 16384.00000  (16384.00000)
     | > grad_norm: 10.82023  (10.45917)
     | > current_lr: 0.00000
     | > step_time: 1.41730  (1.39475)
     | > loader_time: 0.00920  (0.01951)


   --> STEP: 75/203 -- GLOBAL_STEP: 75
     | > loss: 3.67986  (3.68095)
     | > log_mle: 0.75822  (0.75589)
     | > loss_dur: 2.92164  (2.92506)
     | > amp_scaler: 16384.00000  (16384.00000)
     | > grad_norm: 10.71609  (10.55661)
     | > current_lr: 0.00000
     | > step_time: 1.60550  (1.37872)
     | > loader_time: 0.05410  (0.01955)


   --> STEP: 100/203 -- GLOBAL_STEP: 100
     | > loss: 3.70151  (3.68684)
     | > log_mle: 0.75617  (0.75601)
     | > loss_dur: 2.94534  (2.93083)
     | > amp_scaler: 16384.00000  (16384.00000)
     | > grad_norm: 10.82266  (10.60688)
     | > current_lr: 0.00000
     | > step_time: 1.51950  (1.39263)
     | > loader_time: 0.00710  (0.01969)


   --> STEP: 125/203 -- GLOBAL_STEP: 125
     | > loss: 3.69801  (3.68347)
     | > log_mle: 0.75794  (0.75561)
     | > loss_dur: 2.94007  (2.92787)
     | > amp_scaler: 16384.00000  (16384.00000)
     | > grad_norm: 10.69523  (10.62239)
     | > current_lr: 0.00000
     | > step_time: 1.18830  (1.37683)
     | > loader_time: 0.01590  (0.01989)


   --> STEP: 150/203 -- GLOBAL_STEP: 150
     | > loss: 3.69383  (3.68041)
     | > log_mle: 0.75924  (0.75558)
     | > loss_dur: 2.93459  (2.92483)
     | > amp_scaler: 16384.00000  (16384.00000)
     | > grad_norm: 10.58207  (10.62709)
     | > current_lr: 0.00000
     | > step_time: 1.71860  (1.37095)
     | > loader_time: 0.01730  (0.01918)


   --> STEP: 175/203 -- GLOBAL_STEP: 175
     | > loss: 3.65237  (3.68043)
     | > log_mle: 0.75172  (0.75542)
     | > loss_dur: 2.90064  (2.92501)
     | > amp_scaler: 16384.00000  (16384.00000)
     | > grad_norm: 10.60451  (10.63348)
     | > current_lr: 0.00000
     | > step_time: 1.63170  (1.38253)
     | > loader_time: 0.01760  (0.01945)



> DataLoader initialization
| > Tokenizer:
        | > add_blank: False
        | > use_eos_bos: False
        | > use_phonemes: True
        | > phonemizer:
                | > phoneme language: en-us
                | > phoneme backend: gruut
        | > 3 not found characters:
        | > ͡
        | > “
        | > ”
| > Number of instances : 131

   --> STEP: 200/203 -- GLOBAL_STEP: 200
     | > loss: 3.70602  (3.67796)
     | > log_mle: 0.75616  (0.75510)
     | > loss_dur: 2.94985  (2.92285)
     | > amp_scaler: 16384.00000  (16384.00000)
     | > grad_norm: 10.72890  (10.63338)
     | > current_lr: 0.00000
     | > step_time: 0.33760  (1.34546)
     | > loader_time: 0.00410  (0.01908)



> DataLoader initialization
| > Tokenizer:
        | > add_blank: False
        | > use_eos_bos: False
        | > use_phonemes: True
        | > phonemizer:
                | > phoneme language: en-us
                | > phoneme backend: gruut
| > Number of instances : 131
 | > Preprocessing samples
 | > Max text length: 174
 | > Min text length: 20
 | > Avg text length: 100.76335877862596
 |
 | > Max audio length: 222643.0
 | > Min audio length: 34739.0
 | > Avg audio length: 144033.41221374046
 | > Num. instances discarded samples: 0
 | > Batch group size: 0.
 | > Preprocessing samples
 | > Max text length: 174
 | > Min text length: 20
 | > Avg text length: 100.76335877862596
 |
 | > Max audio length: 222643.0
 | > Min audio length: 34739.0
 | > Avg audio length: 144033.41221374046
 | > Num. instances discarded samples: 0
 | > Batch group size: 0.

 > EVALUATION

wɛn ɛni pɹɪznɚ kəmɪtɪd ən əfɛns əɡɛnst ðə kəmjunɪti ɔɹ əɡɛnst ən ɪndɪvɪd͡ʒuəl, hi wəz tɹaɪd baɪ ə kɔɹt ɪn ðə d͡ʒeɪl.
 [!] Character '͡' not found in the vocabulary. Discarding it.
wɪt͡ʃ ɑɹ nɛvɚ dɛstɪnd tə kʌt ðə ɡʌmz, ænd θɹuaʊt ɪts laɪf ðɪs ænəməl ɹɪteɪnz, ɪn ə sɪmɪlɚli ɹudəmɛntɚi kəndɪʃən,
 [!] Character '͡' not found in the vocabulary. Discarding it.
ɡɪnjɑɹd ænd kæləweɪ vjud ðə seɪm laɪnʌp əv fɔɹ mɛn fɹʌm wɪt͡ʃ maɪsəs mɑɹkəm hæd ɚliɚ meɪd hɚ aɪdɛntɪfɪkeɪʃən əv li hɑɹvi ɔzwɔld.
 [!] Character '͡' not found in the vocabulary. Discarding it.
ðæt ðə d͡ʒɑb əv pɹətɛktɪŋ ðə pɹɛzɪdənt hæz nɑt bɪn ænd kænɑt bi ɪksklusɪvli ðə ɹispɑnsɪbɪləti əv ðə sikɹɪt sɚvɪs.
 [!] Character '͡' not found in the vocabulary. Discarding it.
 | > Synthesizing test sentences.

  --> EVAL PERFORMANCE
     | > avg_loader_time: 0.00254 (+0.00000)
     | > avg_loss: 3.76569 (+0.00000)
     | > avg_log_mle: 0.75986 (+0.00000)
     | > avg_loss_dur: 3.00583 (+0.00000)

 > BEST MODEL : /home/xerox/Downloads/TTS/recipes/ljspeech/glow_tts/run-August-15-2022_10+43AM-903a77c1/best_model_203.pth
 ! Run is removed from /home/xerox/Downloads/TTS/recipes/ljspeech/glow_tts/run-August-15-2022_10+43AM-903a77c1
Traceback (most recent call last):
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/trainer.py", line 1533, in fit
    self._fit()
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/trainer.py", line 1527, in _fit
    self.save_best_model()
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/trainer.py", line 1619, in save_best_model
    save_func=self.dashboard_logger.save_model,
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/io.py", line 186, in save_best_model
    fs.copy(checkpoint_path, shortcut_path)
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/fsspec/spec.py", line 881, in copy
    self.cp_file(p1, p2, **kwargs)
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/fsspec/implementations/local.py", line 124, in cp_file
    raise FileNotFoundError(path1)
FileNotFoundError: /home/xerox/Downloads/TTS/recipes/ljspeech/glow_tts/run-August-15-2022_10+43AM-903a77c1/best_model_203.pth
^CTraceback (most recent call last):
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/distribute.py", line 73, in <module>
    distribute()
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/distribute.py", line 59, in distribute
    p.wait()
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/subprocess.py", line 1019, in wait
    return self._wait(timeout=timeout)
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/subprocess.py", line 1653, in _wait
    (pid, sts) = self._try_wait(0)
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/subprocess.py", line 1611, in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
KeyboardInterrupt

**Comment by erogol** [comment]
We need more info to be able to help

**Reply by koayst** [reply]
Please give me 2 - 3 days as the GPUs are used up now.  Once they are available, I will reproduce the problem and post the errors here.  Thanks for the help.

**Reply by erogol** [reply]
Can you try updating the 👟Trainer by `pip install -U trainer`

**Reply by koayst** [reply]
Hi @erogol,

The new trainer solved the issue.  

Thanks a lot.

**Comment by koayst** [comment]
After updating the trainer, the issue is resolved.  Thanks.

**Comment by skshadan** [comment]
Traceback (most recent call last):
  File "/home/ec2-user/tts-sage/recipes/ljspeech/vits_tts/train_vits.py", line 84, in <module>
    trainer = Trainer(
  File "/home/ec2-user/anaconda3/envs/JupyterSystemEnv/lib/python3.10/site-packages/trainer/trainer.py", line 437, in __init__
    self.use_cuda, self.num_gpus = self.setup_training_environment(args=args, config=config, gpu=gpu)
  File "/home/ec2-user/anaconda3/envs/JupyterSystemEnv/lib/python3.10/site-packages/trainer/trainer.py", line 751, in setup_training_environment
    use_cuda, num_gpus = setup_torch_training_env(
  File "/home/ec2-user/anaconda3/envs/JupyterSystemEnv/lib/python3.10/site-packages/trainer/trainer_utils.py", line 100, in setup_torch_training_env
    raise RuntimeError(
RuntimeError:  [!] 8 active GPUs. Define the target GPU by `CUDA_VISIBLE_DEVICES`. For multi-gpu training use `TTS/bin/distribute.py`



i am getting this error!!! how can i use 8 gpus?
